dist/tools/benchmark_udp/main.c

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (C) 2021 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @ingroup     tools
+ * @{
+ *
+ * @file
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/random.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include "list.h"
+#include "kernel_defines.h"
+#include "test_utils/benchmark_udp.h"
+
+#define US_PER_MS   (1000UL)
+#define US_PER_SEC  (1000 * US_PER_MS)
+#define MS_PER_SEC  (1000UL)
+
+typedef struct {
+    list_node_t node;
+    struct sockaddr_in6 addr;
+    struct timeval first_seen;
+    uint32_t seq_no;
+    uint32_t count_tx;
+    uint32_t count_rx;
+    uint32_t count_rt;
+    uint32_t rtt_us;
+} bench_client_t;
+
+static bool one_way;
+static uint32_t cookie;
+static uint32_t delay_us    = 100 * US_PER_MS; /* 100 ms */
+static uint16_t payload_len = 32;
+
+static char addr_str[INET6_ADDRSTRLEN];
+
+static bench_client_t *_find_or_add(list_node_t *head, struct sockaddr_in6 *addr,
+                                    bool *new_node)
+{
+    for (list_node_t* n = head->next; n; n = n->next) {
+        bench_client_t *node = container_of(n, bench_client_t, node);
+
+        if (memcmp(&addr->sin6_addr, &node->addr.sin6_addr, sizeof(addr->sin6_addr)) == 0) {
+            *new_node = false;
+            return node;
+        }
+    }
+
+    inet_ntop(AF_INET6, &addr->sin6_addr, addr_str, INET6_ADDRSTRLEN);
+    printf("adding [%s]:%d\n", addr_str, ntohs(addr->sin6_port));
+    bench_client_t *node = calloc(1, sizeof(bench_client_t));
+    memcpy(&node->addr, addr, sizeof(*addr));
+    list_add(head, &node->node);
+
+    *new_node = true;
+    return node;
+}
+
+static void clrscr(void)
+{
+    printf("\e[1;1H\e[2J");
+}
+
+static uint64_t _tv_diff_msec(struct timeval *a, struct timeval *b)
+{
+    return (a->tv_sec - b->tv_sec) * MS_PER_SEC
+         + (a->tv_usec - b->tv_usec) / US_PER_MS;
+}
+
+static void _print_stats(list_node_t *head, struct timeval *now)
+{
+    static uint8_t max_addr_len;
+
+    printf("host%*s\tbandwith\tnum TX\tnum RX",  max_addr_len - 4, "");
+    if (!one_way) {
+        printf("\t\tnum RT\t\tRTT");
+    }
+    puts("");
+
+    for (list_node_t* n = head->next; n; n = n->next) {
+        bench_client_t *node = container_of(n, bench_client_t, node);
+        uint8_t addr_len;
+
+        inet_ntop(AF_INET6, &node->addr.sin6_addr, addr_str, INET6_ADDRSTRLEN);
+        addr_len = printf("%s", addr_str);
+
+        if (addr_len > max_addr_len) {
+            max_addr_len = addr_len;
+        }
+
+        unsigned bw = (node->count_rx * (payload_len + sizeof(benchmark_msg_ping_t)))
```

Comment:
```suggestion
        unsigned bw = (node->count_rx * (/*bytes_in*/))
```
payload_len may not be the payload length since the client decides about it,
i suggest  storring bytes_in of the last package, that and the bandwidth.

Reply:
if the configuration does not match, we send a config packet and the client will reset it's counters. 

I'd say it's safe to assume that for any valid data, client and server config must & will be in sync.

Reply:
it is about the size of the payload which is limited by the client as far as i can see without further notice to the server